src/libsync/propagatedownloadzsync.cpp

```diff
@@ -0,0 +1,320 @@
+}
+
+bool GETFileZsyncJob::finished()
+{
+    if (reply()->bytesAvailable()) {
+        return false;
+    }
+
+    if (_current < _nrange - 1) {
+        _current++;
+        startCurrentRange();
+        return false;
+    }
+
+    if (!_hasEmittedFinishedSignal) {
+        emit finishedSignal();
+    }
+
+    free(_zbyterange);
+    zsync_end_receive(_zr);
+    zsync_complete(_zs);
+    zsync_end(_zs);
+
+    _hasEmittedFinishedSignal = true;
+
+    deleteLater();
+    return true; // discard
+}
+
+void GETFileZsyncJob::seedFinished(void* zs) {
+    _thread.quit();
+    _thread.wait();
+
+    _zs = (struct zsync_state *)zs;
+    {   /* And print how far we've progressed towards the target file */
+        long long done, total;
+
+        zsync_progress(_zs, &done, &total);
+        qCInfo(lcZsyncGet).nospace() << "Done reading: "
+            << _localpath << " " << fixed << qSetRealNumberPrecision(1) <<  (100.0f * done) / total
+            << "% of target seeded.";
+
+    }
+
+    const char * tmpfile = _localtmppath.toStdString().c_str();
+    if (zsync_rename_file(_zs, tmpfile) != 0) {
+        _errorString = tr("Unable to rename temporary file.");
+        _errorStatus = SyncFileItem::NormalError;
+        qCDebug(lcZsyncGet) << _errorString;
+        if (_zbyterange) free(_zbyterange);
+        zsync_complete(_zs);
+        zsync_end(_zs);
+        emit finishedSignal();
+        return;
+    }
+
+    /* Get a set of byte ranges that we need to complete the target */
+    _zbyterange = zsync_needed_byte_ranges(_zs, &_nrange, 0);
+    if (!_zbyterange) {
+        _errorString = tr("Failed to get zsync byte ranges.");
+        _errorStatus = SyncFileItem::NormalError;
+        qCDebug(lcZsyncGet) << _errorString;
+        if (_zbyterange) free(_zbyterange);
+        zsync_complete(_zs);
+        zsync_end(_zs);
+        emit finishedSignal();
+        return;
+    }
+
+    /* If we have no ranges then we have equal files and we are done */
+    if (_nrange == 0) {
+        if (_zbyterange) free(_zbyterange);
+        zsync_complete(_zs);
+        zsync_end(_zs);
+        emit finishedSignal();
+        return;
+    }
+
+    qCDebug(lcZsyncGet) << "Number of ranges:" << _nrange;
+
+    _zr = zsync_begin_receive(_zs, 0);
+    if(!_zr) {
+        _errorString = tr("Failed to initialize zsync receive structure.");
+        _errorStatus = SyncFileItem::NormalError;
+        qCDebug(lcZsyncGet) << _errorString;
+        if (_zbyterange) free(_zbyterange);
+        zsync_complete(_zs);
+        zsync_end(_zs);
+        emit finishedSignal();
+        return;
+    }
+
+    quint64 zsyncSize = 0;
+    for (int i=0; i<_nrange; i++) {
+        zsyncSize += _zbyterange[(2*i)+1] - _zbyterange[(2*i)];
+    }
+    _propagator->reportFileTotal(*_item, zsyncSize);
+
+    /* start getting bytes for first zsync byte range */
+    startCurrentRange();
+
+    AbstractNetworkJob::start();
```

Comment:
Do this only when you actually start a network job

Reply:
Done.